[front] - feat(skill): BE skill version comparison

JulesBelveze opened the pull request:
## Description
Adds an API endpoint to retrieve the version history of a skill configuration. The endpoint returns all historical versions from the `skill_versions` table along with the current version, sorted by version number descending (newest first). Also adds the `versionAuthorId` field to `SkillConfigurationType` and includes a SWR hook `useSkillConfigurationHistory` for frontend consumption.

## Risk

None not being used yet

## Tests

Locally

## Deploy Plan

- Deploy front as this doesn't require migration

JulesBelveze committed:
[front/pages/api/w/skills/history] - feature: add API endpoint for skill configuration history

 - Implement an API handler to retrieve the history of skill configurations
 - Validate query parameters and ensure the user has the right permissions to access the skill history
 - Return a limited list of skill configurations if a limit is provided in the request

JulesBelveze committed:
[swr] - feature: add hook for fetching skill configuration history

 - Implement a new hook to retrieve the history of skill configurations for a specific skill ID
 - Include support for an optional limit parameter to control the number of history entries fetched
 - Ensure the hook handles loading states, errors, and provides a function to mutate the local data

JulesBelveze committed:
[front/lib/resources/skill] - feature: track author of skill resource versions

 - Include the author id when creating a version of a skill resource to track who made changes
 - Omit version author id for global skill IDs as they are not associated with individual authors

JulesBelveze committed:
[api/assistant/configuration] - feature: implement skill version listing functionality

 - Created a new function to fetch all versions of a skill including the latest version
 - Versions are ordered by descending version numbers with the current version listed first
 - Historical versions are converted to a specific data type ensuring consistent API response format
 - Included permission checks to ensure only authorized users can write to the skill configuration

JulesBelveze committed:
[front/types/api/internal/skill] - feature: add history query schema with limit

 - Introduce a new codec for validating 'limit' within a specific range using io-ts types
 - Extend support for an optional 'limit' in skill configurations history retrieval

JulesBelveze committed:
[api/assistant/configuration] - refactor: improve skill version querying

- Implement a `WhereOptions` type definition for clarity in database queries
- Translate space IDs to their string equivalents for the skill version output

JulesBelveze committed:
[front/lib/api/assistant/configuration] - refactor: reorder import statements in skill.ts

 - Group and relocate the `WhereOptions` import from `sequelize` to enhance import structuring and consistency

JulesBelveze committed:
[front/lib/api/assistant/configuration] - refactor: extract listing skill versions to SkillResource

 - Centralize the logic of listing skill versions to the SkillResource class
 - Remove duplicate code from skill configuration API by leveraging new method in SkillResource

JulesBelveze committed:
[front] - fix: correct loading state for skill configuration history

 - Ensure disabled state is considered when determining the loading state of skill configuration history

JulesBelveze committed:
[front/pages/api/w/[wId]/skills/[sId]/history] - refactor: streamline sId type check in skill history API

 - Replaced redundant type check with isString utility for query parameter sId
 - Removed unnecessary variable assignment for sId after type validation

aubin-tchoi commented on front/lib/resources/skill/skill_resource.ts:
```diff
@@ -577,6 +577,67 @@ export class SkillResource extends BaseResource<SkillConfigurationModel> {
+  static async listVersions(
+    auth: Authenticator,
+    skillId: string
+  ): Promise<SkillConfigurationType[]> {
```
would be nice to return an array of `SkillResource`, and do the toJSON in the handler (closer to the part that answers to the front end)

aubin-tchoi commented on front/lib/resources/skill/skill_resource.ts:
```diff
@@ -577,6 +577,67 @@ export class SkillResource extends BaseResource<SkillConfigurationModel> {
     };
   }
 
+  static async listVersions(
```
I think it would be better to have this be a non-static method, in the endpoint we try to `fetchById`, if we can't then it means we are not allowed to do the fetch so we 404, and then if we can we list the versions for this `skillResource`
wdyt?

JulesBelveze committed:
[front/lib/api/assistant/configuration] - refactor: remove independent version listing function for skill configurations

 - Refactored version listing to be a method of SkillResource instead of a standalone function
 - Removed dependency on workspace existence check in the removed function

[front/lib/models] - feature: add version attribute to SkillVersionModel

 - Introduced a version field in SkillVersionModel to track version numbers

[front/lib/resources/skill] - refactor: optimize version listing and conversion to SkillResource instances

 - Redesigned version listing to sort versions application-side and instantiate sorted SkillResource objects
 - Altered listVersions return type from SkillConfigurationType[] to an array of SkillResource instances

[front/pages/api/w/[wId]/skills/[sId]/history] - fix: correct API behavior for fetching skill history

 - Fixed skill fetch conditional to disallow access for unauthorized users
 - Refactored handler to work directly with SkillResource for listing versions and conversion to JSON
 - Updated error handling to return 404 when skill cannot be fetched or user lacks permissions

JulesBelveze commented on front/pages/api/w/[wId]/skills/[sId]/history/index.ts:
```diff
@@ -0,0 +1,94 @@
+  // Check that user has access to this skill
+  const skill = await SkillResource.fetchById(auth, sId);
+
+  if (!skill || (!skill.canWrite(auth) && !auth.isAdmin())) {
```
I'm always scared haha - but I think you right

aubin-tchoi commented on front/lib/swr/skill_configurations.ts:
```diff
@@ -203,3 +204,34 @@ export function useRestoreSkillConfiguration({
+  disabled,
+}: {
+  owner: LightWorkspaceType;
+  skillConfigurationId: string | null;
```
I think we will probably have a `SkillConfigurationType`, which we can use here instead of only the sId

aubin-tchoi reviewed (approved):
LGTM ✨

JulesBelveze commented on front/lib/resources/skill/skill_resource.ts:
```diff
@@ -577,6 +577,65 @@ export class SkillResource extends BaseResource<SkillConfigurationModel> {
+            id: -1,
+            workspaceId: workspace.id,
+            skillConfigurationId: this.id,
+            mcpServerViewId: mcpServerId,
```
Yeah probably easier to understand - will do in a follow up PR

JulesBelveze committed:
[front/lib/resources] - refactor: use provided model in SkillResource

 - Use `this.model` directly instead of `SkillResource.model` to ensure correct model instance is used

[front/lib/swr] - refactor: streamline skill configuration history API

 - Replace `skillConfigurationId` with `skillConfiguration` object to include additional configuration properties in requests
 - Construct API endpoint using `skillConfiguration`'s `sId` property for history fetch calls

[front/lib/resources] - fix: remove fallback to current date in `createdAt` and `updatedAt`

 - Ensure `createdAt` and `updatedAt` fields are assigned directly from `versionModel` without fallback values
